### DISCUSSION TITLE
Since Laravel 9: Symfony mailer: Unable to connect with STARTTLS - verify_peer

### DISCUSSION THREAD

**Discussion by michabbb** [laravel/framework#41385] [General]
hi there,

i recently upgraded to laravel 9 and because of the Symfony mailer I had problems sending emails, I never had before.
after debugging the Symfony mailer, I want to explain what's different and maybe someone who knows how to contribute
to the laravel docu is able to make a suggestion there because things in the "conf/mail.php" have changed, but the docu
does not mirror that, or at least I didn´t see that.

with laravel <=8, connecting a mail server via IP (not hostname) didn´t run into any problems, when it comes to checking the certificate, maybe because the cert never got verified I don´t know.
with laravel >=9, the cert gets verified (maybe by default) and with that, I ran into the error:

```
Unable to connect with STARTTLS: stream_socket_enable_crypto(): 
Peer certificate CN=`some.domain-name.de' did not match 
expected CN=`XX.XXX.XXX.XXX'  
```

so first: the error is legit, no questions ask 😏 
so in case you are not in control of the hostname or IP, you may be forced to disable that check.
because I didn´t know how to do this, google let me find something like this, **but that only applies to laravel <=8**:

```php
'smtp'         => [
            'transport'  => 'smtp',
            'host'       => env('MAIL_HOST', 'smtp.mailgun.org'),
            'port'       => env('MAIL_PORT', 587),
            'encryption' => env('MAIL_ENCRYPTION', 'tls'),
            'username'   => env('MAIL_USERNAME'),
            'password'   => env('MAIL_PASSWORD'),
            'timeout'    => null,
            'auth_mode'  => null,
            'stream'     => [
                'ssl' => [
                    'allow_self_signed' => true,
                    'verify_peer'       => false,   // <-------------- this is the important part 
                    'verify_peer_name'  => false,
                ],
            ],
        ],
```

but what I am missing in the latest documentation and in the upgrade guide is the information,
that **with laravel 9 it now has to look like this**:

```php
'smtp'         => [
            'transport'  => 'smtp',
            'host'       => env('MAIL_HOST', 'smtp.mailgun.org'),
            'port'       => env('MAIL_PORT', 587),
            'encryption' => env('MAIL_ENCRYPTION', 'tls'),
            'username'   => env('MAIL_USERNAME'),
            'password'   => env('MAIL_PASSWORD'),
            'timeout'    => null,
            'auth_mode'  => null,
            'verify_peer'       => false,   // <------ IT HAS TO BE HERE
        ],
```

and with "verify_peer" _at this position_, Symfony mailer now uses that option and disables any cert check.

i would be happy if someone could help to add this to the official documentation to avoid any further problems for other dev that have to deal with "disabling cert verification".

thanks!

**Comment by erikn69** [comment]
Thanks for the information
So are not  `allow_self_signed` and `verify_peer_name` needed?

**Reply by michabbb** [reply]
I didn't check these two options, my problem just was to ignore the certificate's peer, but i wouldn't be surprised if there are also changes about the other two (you mentioned)

**Comment by BerendSpigt** [comment]
I Agree this should be in the documentation (it is mentioned in the upgrade guide).

But the best way to handle this, would be to add this to the default mail.php conf file so it would look like this:
```php
    'mailers' => [
        'smtp' => [
            'transport' => 'smtp',
            'host' => env('MAIL_HOST', 'smtp.mailgun.org'),
            'port' => env('MAIL_PORT', 587),
            'encryption' => env('MAIL_ENCRYPTION', 'tls'),
            'username' => env('MAIL_USERNAME'),
            'password' => env('MAIL_PASSWORD'),
            'timeout' => null,
            'verify_peer' => env('MAIL_VERIFY_PEER', true),
        ],
```

**Reply by iafilin** [reply]
I upgraded Laravel 5.7 to version 10. As far as I can tell from the code and personal experience, this option only works on configuration level zero.
![image](https://github.com/laravel/framework/assets/26049079/d4f48bf9-de64-4563-8ccf-9b6d6b7e9d48)

**Reply by shohelrana** [reply]
I have upgraded laravel from 5.4 to 11.  This solution is working for me. Thanks

**Comment by 334427** [comment]
Thank you for posting this. It solved my problems for the day!

**Reply by mdsuzayethossan** [reply]
Are you available now? I can't solve this, please help me.

**Comment by Alvalens** [comment]
Thank you, I've been searching solutions for this problem and nothing work, and finally found it here

**Comment by couto862** [comment]
Thank you group for the information. I added the line 'verify_peer => false' in config/mail.php and resolved!

**Comment by amitsolanki1** [comment]
Does it also solve my problem?
I'm using laravel 7.30.6 and yeah it's too old version.

I got this error first time on my production.

**Comment by ahmeti** [comment]
If it still does not work, a new parameter has been added for Symfony Mailer. See also [auto_tls](https://github.com/symfony/mailer/commit/878779807b03d149f503ad1c84bf0a22a5e87c58)

```php
'smtp' => [
    'transport'   => 'smtp',
    'host'        => env('MAIL_HOST', 'smtp.mailgun.org'),
    'port'        => env('MAIL_PORT', 587),
    'encryption'  => env('MAIL_ENCRYPTION', 'tls'),
    'username'    => env('MAIL_USERNAME'),
    'password'    => env('MAIL_PASSWORD'),
    'timeout'     => null,
    'auth_mode'   => null,
    'verify_peer' => false,
    'auto_tls'    => false, // Important this line
]
```

**Reply by amitsolanki1** [reply]
Thanks for your response.

**Reply by Hekkydev** [reply]
verry helpfull

**Comment by toto975** [comment]
In symfony\mailer\Transport\Smtp\Stream\SocketStream.php, replace

`$options['ssl']['crypto_method'] ??= \STREAM_CRYPTO_METHOD_TLS_CLIENT | \STREAM_CRYPTO_METHOD_TLSv1_2_CLIENT | \STREAM_CRYPTO_METHOD_TLSv1_1_CLIENT;`

with

`$options['ssl']['crypto_method'] ??= \STREAM_CRYPTO_METHOD_TLSv1_2_CLIENT;`

It would be great if Symfony adds the crypto_method parameter in the mailer config.
